File: impl/go/date_test.go
package jed

// Cross-check: the Go date parser/renderer reproduces the byte-exact vectors in
// spec/encoding/dates.toml (CLAUDE.md §8) — identical to the Rust/TS cores. Reuses the
// timestamp vector scanner (readTimestampCases / unquote / parseInlineFields). Test-only.

import (
	"strconv"
	"testing"
)

func TestDateVectors(t *testing.T) {
	cases := readTimestampCases(t, specPath(t, "encoding/dates.toml"))
	if len(cases) == 0 {
		t.Fatal("no date vectors parsed")
	}
	for _, c := range cases {
		if c.typ != "date" {
			t.Fatalf("unexpected vector type %q", c.typ)
		}
		switch c.section {
		case "parse":
			in := unquote(c.fields["input"])
			want, _ := strconv.ParseInt(c.fields["days"], 10, 64)
			got, err := ParseDate(in)
			switch {
			case err != nil:
				t.Errorf("parse %q: unexpected error %v", in, err)
			case int64(got) != want:
				t.Errorf("parse %q = %d, want %d", in, got, want)
			}
		case "parse_error":
			in := unquote(c.fields["input"])
			want := unquote(c.fields["error"])
			_, err := ParseDate(in)
			ee, ok := err.(*EngineError)
			if !ok {
				t.Errorf("parse %q: expected error %s, got %v", in, want, err)
			} else if ee.Code() != want {
				t.Errorf("parse %q: error %s, want %s", in, ee.Code(), want)
			}
		case "render":
			d, _ := strconv.ParseInt(c.fields["days"], 10, 64)
			want := unquote(c.fields["text"])
			if got := RenderDate(int32(d)); got != want {
				t.Errorf("render %d = %q, want %q", d, got, want)
			}
		}
	}
}

File: impl/ts/tests/date.test.ts
// Cross-check: the TS date parser/renderer reproduces the byte-exact vectors in
// spec/encoding/dates.toml (CLAUDE.md §8) — identical to the Rust/Go cores. The day count is a
// bigint (the core's uniform-integer discipline). Reuses the timestamp vector scanner.

import assert from "node:assert/strict";
import { test } from "node:test";
import { parseDate, renderDate } from "../src/date.ts";
import { EngineError } from "../src/errors.ts";
import { readTimestampCases, specPath } from "./tomlmini.ts";

test("date vectors (parse/render byte-identical to Rust/Go)", () => {
  const cases = readTimestampCases(specPath("encoding/dates.toml"));
  assert.ok(cases.length > 0, "no date vectors parsed");
  for (const c of cases) {
    assert.equal(c.typ, "date", "unexpected vector type");
    if (c.section === "parse") {
      const got = parseDate(c.fields.input!);
      assert.equal(got, BigInt(c.fields.days!), `parse ${JSON.stringify(c.fields.input)}`);
    } else if (c.section === "parse_error") {
      let code = "";
      try {
        parseDate(c.fields.input!);
      } catch (e) {
        if (e instanceof EngineError) code = e.code();
        else throw e;
      }
      assert.equal(code, c.fields.error, `parse ${JSON.stringify(c.fields.input)} error`);
    } else {
      const got = renderDate(BigInt(c.fields.days!));
      assert.equal(got, c.fields.text, `render ${c.fields.days}`);
    }
  }
});
